### PR TITLE
Support empty regex code sets

### PR DESCRIPTION
**Pull request by tjpalmer**
- Fix temperlang/temper-regex-parser#26
- Support both `/[^]/` (match anything, including newlines) and `/[]/` (match nothing)

### CODE REVIEW

File: be-csharp/src/commonMain/resources/lang/temper/be/csharp/std/Regex/RegexSupport.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ internal static IRegexNode AdjustCodeSet(
             RegexRefs regexRefs
         )
         {
-            if (!codeSet.Negated)
+            if (!codeSet.Negated || codeSet.Items.Count == 0)
```
Comment (tjpalmer) [inline be-csharp/src/commonMain/resources/lang/temper/be/csharp/std/Regex/RegexSupport.cs:43]: Customizing `/[^]/` is elsewhere, so don't go crazy on it here.

File: frontend/src/commonMain/resources/std/regex/regex.temper.md [hunk 2]
```diff
@@ -567,14 +567,26 @@ class RegexFormatter {
...
-          pushCodeSetItem(adjusted.items[i]);
+        if (adjusted.items.isEmpty) {
+          // Many regex engines don't like empty code sets.
+          if (adjusted.negated) {
+            // Match anything.
+            out.append(raw"[\s\S]");
+          } else {
+            // Match nothing.
+            out.append("(?:$.)");
+          }
```
Comment (tjpalmer) [inline frontend/src/commonMain/resources/std/regex/regex.temper.md:578]: Here's where we customize empty code sets. Both of these seem to work across all backends right now. Interp doesn't use this code path, though.

File: functional-test-suite/src/commonMain/resources/regex/match/match.temper.md [hunk 3]
```diff
@@ -215,6 +215,26 @@ negated basic true: true
...
+
+And this should match nothing.
+
+    console.log(/[]+/.find(simplerLines).full.value orelse "-");
+
+```log
+It was the best of times.
+It was the worst of times.
+-
+```
```
Comment (tjpalmer) [inline functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:244]: Match everything and match nothing tests.

File: regex/src/commonMain/kotlin/lang/temper/regex/RegexFormat.kt [hunk 4]
```diff
@@ -287,12 +287,24 @@ private fun PatternFormatContext.formatUnwrapped(codeRange: CodeRange) {
...
+                    if (adjusted.negated) {
+                        out.append('^')
+                    }
+                    adjusted.items.forEach { formatCodeSetItem(it) }
+                    out.append(']')
+                }
             }
-            adjusted.items.forEach { formatCodeSetItem(it) }
-            out.append(']')
         }
```
Comment (tjpalmer) [inline regex/src/commonMain/kotlin/lang/temper/regex/RegexFormat.kt:308]: This is where the logic for interp is, so duplicated here.

File: regex/src/commonTest/kotlin/lang/temper/regex/RegexMatchTest.kt [hunk 5]
```diff
@@ -78,6 +78,7 @@ class RegexMatchTest {
             "moreBoundaryUgly" to
                 RegexCheck(true, RegexMatch("""(?!$word)ø${wordBoundary}8""", "ø8")),
             "digitsNonAsciiUgly" to RegexCheck(false, RegexMatch("""[0-9]+""", "\u0662")),
+            "matchAll" to RegexCheck(true, RegexMatch("""[\s\S]+""", "a\nb")),
```
Comment (tjpalmer) [inline regex/src/commonTest/kotlin/lang/temper/regex/RegexMatchTest.kt:81]: I didn't realize where the problem was in dotnet for a while, so I was testing this here. Decided to leave it.

File: regex/src/commonTest/kotlin/lang/temper/regex/RegexMatchTest.kt [hunk 6]
```diff
@@ -202,7 +203,7 @@ internal fun checkMatchDotnet(checks: Iterable<RegexCheck>) {
             List.map
                 (fun pair ->
                     try
-                        Nullable<bool>(Regex.IsMatch(pair.matchText, pair.pattern))
+                        Nullable<bool>(Regex.IsMatch(pair.matchText, "^" + pair.pattern + "$"))
```
Comment (tjpalmer) [inline regex/src/commonTest/kotlin/lang/temper/regex/RegexMatchTest.kt:206]: Realized I wasn't testing full regexes, but I think it was implied to do so. All the current tests pass either way, but I could construct cases where it made a difference.

File: frontend/src/commonMain/resources/std/regex/regex.temper.md (hunk 2 above)
Comment (mikesamuel) [reply frontend/src/commonMain/resources/std/regex/regex.temper.md:578]: Does the difference between `$` and `\z` matter here?

I guess `.` doesn't match anything allowed by `.` on perlre.

File: functional-test-suite/src/commonMain/resources/regex/match/match.temper.md [hunk 7]
```diff
@@ -215,6 +215,26 @@ negated basic true: true
 negated basic false: false
 ```
 
+### Empty sets
+
+This should match everything, including newlines.
+
+    let simplerLines = """
+      "It was the best of times.
+      "It was the worst of times.
```
Comment (mikesamuel) [inline functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:224]: It was the best of lines.
It was the worst of lines.

File: functional-test-suite/src/commonMain/resources/regex/match/match.temper.md [hunk 8]
```diff
@@ -215,6 +215,26 @@ negated basic true: true
...
+    ;
+    console.log(/[^]+/.find(simplerLines).full.value);
+
+And this should match nothing.
+
+    console.log(/[]+/.find(simplerLines).full.value orelse "-");
```
Comment (mikesamuel) [inline functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:230]: I guess the semantics of []* are that it still matches the empty string.

File: frontend/src/commonMain/resources/std/regex/regex.temper.md (hunk 2 above)
Comment (tjpalmer) [reply frontend/src/commonMain/resources/std/regex/regex.temper.md:578]: > Does the difference between `$` and `\z` matter here?

I think `$` only matches end of string on our current configurations. Part of the reason to test a multiline string. I think it's safe. Do you have a particular concern?

File: functional-test-suite/src/commonMain/resources/regex/match/match.temper.md (hunk 8 above)
Comment (tjpalmer) [reply functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:230]: > I guess the semantics of []* are that it still matches the empty string.

Just tested it, and seems it works that way everywhere except Lua. 😭

I can add a commit with this case added but commented out for future reference, if you want. It seems edge case enough to me that we can defer it.

Comment (tjpalmer) [reply functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:230]: I added a PR with the commented-out case.

File: functional-test-suite/src/commonMain/resources/regex/match/match.temper.md [hunk 9]
```diff
@@ -215,6 +215,34 @@ negated basic true: true
...
+
+But this should match a different kind of nothing, the empty string.
+
+```temper inert
+    // Ignore this case, because it fails on Lua.
+    // If enabled, we need a blank line added to the next log block.
+    console.log(/[]*/.find(simplerLines).full.value);
+```
```
Comment (tjpalmer) [inline functional-test-suite/src/commonMain/resources/regex/match/match.temper.md:238]: Added because of [this comment](https://github.com/temperlang/temper/pull/347#discussion_r2829935549).
